content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md

```diff
@@ -0,0 +1,267 @@
+
+```diff
+diff --git a/pkg/gce-pd-csi-driver/gce-pd-driver.go b/pkg/gce-pd-csi-driver/gce-pd-driver.go
+index 45903f3..0d7ea26 100644
+--- a/pkg/gce-pd-csi-driver/gce-pd-driver.go
++++ b/pkg/gce-pd-csi-driver/gce-pd-driver.go
+@@ -56,6 +56,8 @@ func (gceDriver *GCEDriver) SetupGCEDriver(name, vendorVersion string, extraVolu
+                csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+                csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY,
+                csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER,
++               csi.VolumeCapability_AccessMode_SINGLE_NODE_SINGLE_WRITER,
++               csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER,
+        }
+        gceDriver.AddVolumeCapabilityAccessModes(vcam)
+        csc := []csi.ControllerServiceCapability_RPC_Type{
+@@ -67,12 +69,14 @@ func (gceDriver *GCEDriver) SetupGCEDriver(name, vendorVersion string, extraVolu
+                csi.ControllerServiceCapability_RPC_EXPAND_VOLUME,
+                csi.ControllerServiceCapability_RPC_LIST_VOLUMES,
+                csi.ControllerServiceCapability_RPC_LIST_VOLUMES_PUBLISHED_NODES,
++               csi.ControllerServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
+        }
+        gceDriver.AddControllerServiceCapabilities(csc)
+        ns := []csi.NodeServiceCapability_RPC_Type{
+                csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
+                csi.NodeServiceCapability_RPC_EXPAND_VOLUME,
+                csi.NodeServiceCapability_RPC_GET_VOLUME_STATS,
++               csi.NodeServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
+        }
+        gceDriver.AddNodeServiceCapabilities(ns)
+```
+
+### Implement `NodePublishVolume` behavior
+
+The CSI spec outlines expected behavior for the `NodePublishVolume` RPC when called more than once for the same volume but with different arguments (like the target path).
+Please refer to [the second table in the NodePublishVolume section of the CSI spec](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodepublishvolume) for more details on expected behavior when implementing in your driver.
+
+### Update your CSI sidecars (optional)
+
+When deploying your CSI drivers, you may update your CSI sidecars (mainly the [external-provisioner](https://github.com/kubernetes-csi/external-provisioner), [external-attacher](https://github.com/kubernetes-csi/external-attacher), and [external-resizer](https://github.com/kubernetes-csi/external-resizer)) to versions that depend on CSI spec v1.5.0+ and the Kubernetes v1.22 API.
+
+Note that this is not required, as the ReadWriteOncePod access mode will work out of the box without any changes to your driver.
+However, if you would like to leverage the new CSI access modes and capabilities added to your driver, this step is recommended.
+
+## What’s next?
+
+As part of the beta graduation for this feature, support will be added for pod preemption in the scheduler.
```

Comment:
- I'd write “we plan to” for statements like this (exception: SIG Architecture have signed off on a commitment that it will happen this way)
- the scheduler already supports preemption; I've reworded to clarify what would be added
```suggestion
As part of the beta graduation for this feature, we plan to update the Kubenetes scheduler to support pod preemption in relation to ReadWriteOncePod storage.
```

(also see separate comment about making sure the reader is clear who “we” refers to).